File: kotlin/hexadecimal_to_decimal_converter/hexadecimal_to_decimal_converter_9.kt
/**
 * This Kotlin program is designed to convert hexadecimal numbers to their decimal equivalents.
 * Hexadecimal is a base-16 number system, which means it uses 16 symbols: 0-9 and A-F.
 * This program will take a hexadecimal number as input and output its decimal representation.
 * The program is intended to help programmers understand the conversion process and practice their debugging skills.
 */

fun main() {
    // Prompt the user for a hexadecimal number
    println("Enter a hexadecimal number:")
    val hexInput = readLine() ?: ""

    // Call the conversion function and store the result in a variable
    val decimalOutput = convertHexToDecimal(hexInput)

    // Display the result to the user
    println("The decimal equivalent of $hexInput is $decimalOutput")
}

/**
 * This function converts a hexadecimal string to its decimal equivalent.
 * It iterates over each character in the hexadecimal string, calculates its decimal value,
 * and accumulates the result.
 *
 * @param hex The hexadecimal string to be converted.
 * @return The decimal equivalent of the hexadecimal string.
 */
fun convertHexToDecimal(hex: String): Int {
    // Initialize the result variable to store the final decimal value
    var result = 0

    // Initialize a variable to keep track of the current power of 16
    var power = 1

    // Iterate over the hexadecimal string from right to left
    for (i in hex.length - 1 downTo 0) {
        // Get the current character
        val c = hex[i]

        // Determine the decimal value of the current character
        val decimalValue = when (c) {
            in '0'..'9' -> c - '0'
            in 'A'..'F' -> c - 'A' + 10
            in 'a'..'f' -> c - 'a' + 10
            else -> 0 // This case should never happen for valid hexadecimal input
        }

        // Accumulate the result by adding the current decimal value multiplied by the current power of 16
        result += decimalValue * power

        // Update the power of 16 for the next iteration
        power *= 16
    }

    // Return the final decimal result
    return

File: ruby/hexadecimal_to_decimal_converter/hexadecimal_to_decimal_converter_9.rb
# This program is designed to convert hexadecimal numbers to their decimal equivalents.
# Hexadecimal numbers are base-16 numbers, which means they use 16 different symbols: 0-9 and A-F.
# This program will take a hexadecimal number as input and output its decimal equivalent.
# The program will demonstrate the conversion process in a clear and detailed manner.

def hex_to_dec(hex)
    # Initialize the decimal value to zero
    decimal = 0
    
    # Define a hash map to store hexadecimal characters and their corresponding decimal values
    hex_map = {
        '0' => 0, '1' => 1, '2' => 2, '3' => 3, '4' => 4, '5' => 5, '6' => 6, '7' => 7, '8' => 8, '9' => 9,
        'A' => 10, 'B' => 11, 'C' => 12, 'D' => 13, 'E' => 14, 'F' => 15
    }
    
    # Reverse the hexadecimal string to process from least significant digit to most significant digit
    hex.reverse.each_char.with_index do |char, index|
        # Convert the hexadecimal character to its decimal value using the hash map
        value = hex_map[char.upcase]
        
        # Calculate the contribution of the current digit to the overall decimal value
        decimal += value * (16 ** index)
    end
    
    # Return the final decimal value
    decimal
end

# Function to get user input and display the result
def main
    # Prompt the user to enter a hexadecimal number
    puts "Please enter a hexadecimal number:"
    hex_number = gets.chomp
    
    # Convert the hexadecimal number to decimal
    decimal_number = hex_to_dec(hex_number)
    
    # Display the result to the user
    puts "The decimal equivalent of #{hex_number} is #{decimal_number}."
end

# Call the main function to execute the program
main

